File: generic-functions-test.cpp
#include "generic-functions-test.h"

using namespace js;

template<typename T>
T identity(T value) {
    return value;
}

template<typename T>
js::number getLength(T item) {
    return item.length();
}

template<typename T, typename U>
js::any swap(T a, U b) {
    return js::array<js::any>{b, a};
}

template<typename T>
js::array<T> createArray(js::number size, T defaultValue) {
    return std::make_shared<js::array>(size).fill(defaultValue);
}

Container::Container(T initialValue) {
    {
            this->value = initialValue;
    }
}

T Container::getValue() {
    {
            return this->value;
    }
}

void Container::setValue(T newValue) {
    {
            this->value = newValue;
    }
}

Container Container::transform(std::function<U(T)> fn) {
    {
            return std::make_shared<Container>(fn(this->value));
    }
}

template<typename T, typename U>
Pair createPair(T first, U second) {
    return []() {
          js::object obj_temp_0;
          obj_temp_0.set("first", first);
          obj_temp_0.set("second", second);
          return js::any(obj_temp_0);
        }();
}

void testGenericFunctions() {
    const js::any stringIdentity = identity("hello"_S);
    const js::any numberIdentity = identity(js::number(42));
    const js::any boolIdentity = identity(true);
    const js::any strLength = getLength("hello world"_S);
    const js::any arrayLength = getLength(js::array<js::number>{js::number(1), js::number(2), js::number(3), js::number(4), js::number(5)});
    const js::any swapped = swap(js::number(10), "hello"_S);
    const js::any stringArray = createArray(js::number(3), "default"_S);
    const js::any numberArray = createArray(js::number(3), js::number(0));
    const std::shared_ptr<Container> stringContainer = std::make_shared<Container>("initial"_S);
    const js::any transformedContainer = stringContainer->transform([](auto s) -> auto { return s.length(); });
    const js::any pair = createPair("key"_S, js::number(123));
    js::console.log("Generic tests:"_S, []() {
          js::object obj_temp_1;
          obj_temp_1.set("stringIdentity", stringIdentity);
          obj_temp_1.set("numberIdentity", numberIdentity);
          obj_temp_1.set("boolIdentity", boolIdentity);
          obj_temp_1.set("strLength", strLength);
          obj_temp_1.set("arrayLength", arrayLength);
          obj_temp_1.set("swapped", swapped);
          obj_temp_1.set("stringArray", stringArray);
          obj_temp_1.set("numberArray", numberArray);
          obj_temp_1.set("stringValue", stringContainer->getValue());
          obj_temp_1.set("transformedValue", transformedContainer->getValue());
          obj_temp_1.set("pair", pair);
          return js::any(obj_temp_1);
        }());
}

// Entry point
void Main() {
    testGenericFunctions();
}

int main(int /*argc*/, char** /*argv*/) {
    Main();
    return 0;
}

File: tests/integration/generic-functions-test.ts
// Generic functions test

// Basic generic function
function identity<T>(value: T): T {
  return value;
}

// Generic function with constraints
function getLength<T extends { length: number }>(item: T): number {
  return item.length;
}

// Multiple type parameters
function swap<T, U>(a: T, b: U): [U, T] {
  return [b, a];
}

// Generic function with default type parameter
function createArray<T = string>(size: number, defaultValue: T): T[] {
  return new Array(size).fill(defaultValue);
}

// Generic class with methods
class Container<T> {
  private value: T;

  constructor(initialValue: T) {
    this.value = initialValue;
  }

  getValue(): T {
    return this.value;
  }

  setValue(newValue: T): void {
    this.value = newValue;
  }

  transform<U>(fn: (value: T) => U): Container<U> {
    return new Container(fn(this.value));
  }
}

// Generic interface
interface Pair<T, U> {
  first: T;
  second: U;
}

function createPair<T, U>(first: T, second: U): Pair<T, U> {
  return { first, second };
}

function testGenericFunctions(): void {
  // Test basic generic function
  const stringIdentity = identity("hello");
  const numberIdentity = identity(42);
  const boolIdentity = identity(true);

  // Test constraint generic
  const strLength = getLength("hello world");
  const arrayLength = getLength([1, 2, 3, 4, 5]);

  // Test multiple parameters
  const swapped = swap(10, "hello");

  // Test default type parameter
  const stringArray = createArray(3, "default");
  const numberArray = createArray<number>(3, 0);

  // Test generic class
  const stringContainer = new Container("initial");
  const transformedContainer = stringContainer.transform((s) => s.length);

  // Test generic interface
  const pair = createPair("key", 123);

  console.log("Generic tests:", {
    stringIdentity,
    numberIdentity,
    boolIdentity,
    strLength,
    arrayLength,
    swapped,
    stringArray,
    numberArray,
    stringValue: stringContainer.getValue(),
    transformedValue: transformedContainer.getValue(),
    pair,
  });
}

testGenericFunctions();
